### PR TITLE
Implement merge operator, bugfixes for flatMap

### PR DESCRIPTION
**Pull request by dymk**
Implements the merge operator for Flowable, which converts a Flowable<Flowable<T>> into a Flowable<T>. It's implemented as a specialization of `flatMap`. 

Also included are bugfixes for flatMap, in which flatMap would terminate its subscriber too early, once all of this sub-streams had completed (but possibly still had outstanding elements to be consumed).

### CODE REVIEW

File: yarpl/include/yarpl/flowable/FlowableOperator.h [hunk 1]
```diff
@@ -542,13 +539,13 @@ class FlatMapOperator : public FlowableOperator<T, R, FlatMapOperator<T, R>> {
     void drainImpl() {
       // phase 1: clear out terminated subscribers
       {
-        auto clearList = [](auto& list) {
+        SubscriberList trash;
+        auto clearList = [&trash](auto& list) {
```
Comment (phoad) [inline yarpl/include/yarpl/flowable/FlowableOperator.h]: nit: if you make the 'trash' another input to the function, then you might declare that variable when the lambda gets used.

File: yarpl/test/FlowableFlatMapTest.cpp [hunk 2]
```diff
@@ -271,5 +271,44 @@ TEST(FlowableFlatMapTest, MultithreadedLargeAmount) {
   p2->evb.stop();
 }
 
+TEST(FlowableFlatMapTest, MergeTest) {
+  auto sub = yarpl::make_ref<TestSubscriber<std::string>>(0);
+
+  auto p1 = Flowables::justN<std::string>({"foo", "bar"});
```
Comment (phoad) [inline yarpl/test/FlowableFlatMapTest.cpp]: It is possible to write a test like this
 - One flowable from Thread A
 - One flowable from Thread B
 - They are merged into another flowable which lives in Thread C
So it will be a total merge of the flowables from different threads into a single thread.

Comment (dymk) [reply yarpl/test/FlowableFlatMapTest.cpp]: Tests in this compilation unit do exercise multiple independent threads pushing into flatMap (and therefore merge, as merge is just a very light wrapper around flatMap)

File: yarpl/include/yarpl/flowable/FlowableOperator.h (hunk 1 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/FlowableOperator.h]: I'm not sure I follow, you mean don't capture trash, but pass it in as another parameter?

Comment (phoad) [reply yarpl/include/yarpl/flowable/FlowableOperator.h]: nit: would it be better not not?

File: yarpl/include/yarpl/flowable/Flowable.h [hunk 3]
```diff
@@ -20,6 +20,19 @@
...
 
+template <typename T>
+class Flowable;
+
+namespace detail {
+
+template <typename T>
+struct is_t_flowable : std::false_type {};
+
+template <typename R>
```
Comment (phoad) [inline yarpl/include/yarpl/flowable/Flowable.h]: Do we need all these or do we make the code more complex, do we have hard requirement for these?

template <typename T> , should we use T instead of R?

File: yarpl/include/yarpl/flowable/Flowable.h [hunk 4]
```diff
@@ -100,6 +113,16 @@ class Flowable : public virtual Refcounted, public yarpl::enable_get_ref {
...
   Reference<Flowable<T>> observeOn(folly::Executor&);
 
+  template <typename Q>
+  using enableWrapRef =
+      typename std::enable_if<detail::is_t_flowable<Q>::value, Q>::type;
+
+  template <typename Q = T>
+  enableWrapRef<Q> merge() {
+    folly::Function<T(T)> func = [](auto f) { return std::move(f); };
+    return this->flatMap(std::move(func));
```
Comment (phoad) [inline yarpl/include/yarpl/flowable/Flowable.h]: Would it be possible to just copy paste the 'lambda' part into this. compiler might generate an optimized version instead of folly::Function..

this->flatMap([](auto f) { return std::move(f); });

File: yarpl/include/yarpl/flowable/FlowableOperator.h (hunk 1 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/FlowableOperator.h]: I think it won't matter; in any optimized build the lambda will be inlined

File: yarpl/include/yarpl/flowable/Flowable.h [hunk 5]
```diff
@@ -100,6 +113,16 @@ class Flowable : public virtual Refcounted, public yarpl::enable_get_ref {
 
   Reference<Flowable<T>> observeOn(folly::Executor&);
 
+  template <typename Q>
+  using enableWrapRef =
```
Comment (phoad) [inline yarpl/include/yarpl/flowable/Flowable.h]: Do you use this more than once? Do we need to define it at all?

File: yarpl/include/yarpl/flowable/Flowable.h [hunk 6]
```diff
@@ -20,6 +20,19 @@
...
+template <typename T>
+class Flowable;
+
+namespace detail {
+
+template <typename T>
+struct is_t_flowable : std::false_type {};
+
+template <typename R>
+struct is_t_flowable<Reference<Flowable<R>>> : std::true_type {};
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/flowable/Flowable.h]: Nit: I believe the convention is `is_flowable` for the struct type and `is_flowable_v` for the value type.

Thinking of `std::is_signed` and `std::is_signed_v`.

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 3 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/Flowable.h]: Well the R references a different variable than T; T is `is_t_flowable<T>`, R is the template parameter that `T` is parameterized over in the case that `T` is a `Reference<Flowable<R>>`

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 5 above)
Comment (alexmalyshev) [reply yarpl/include/yarpl/flowable/Flowable.h]: Unrelated, but I prefer `static_assert` always to `enable_if`.

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 4 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/Flowable.h]: A limitation of flatMap right now is that it can't infer return type from the type of the lambda (templates are... hard. I can work on this more in another diff). It's held in flatMap as a folly::Function, so there won't be a perf hit.

Comment (alexmalyshev) [reply yarpl/include/yarpl/flowable/Flowable.h]: +1, I think it'd make things a bit cleaner.

Doesn't really matter, but is the `std::move` necessary here?  I can never remember what the rules for RVO are, and which C++ standard made it mandatory.

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 5 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/Flowable.h]: It's to help with readability. IMO a complex inline `enable_if` is pretty illegible as the return type for `merge`.

File: yarpl/include/yarpl/flowable/FlowableOperator.h [hunk 7]
```diff
@@ -620,10 +627,37 @@ class FlatMapOperator : public FlowableOperator<T, R, FlatMapOperator<T, R>> {
...
+
+          // Subscribers might call onNext and then terminate; delay
+          // removing its liveSubscriber reference until we've delivered
+          // its element to the downstream subscriber and dropped its
+          // synchronized reference to `r`, as dropping the flatMapSubscription_
+          // reference may invoke its destructor
+          if (r->isTerminated) {
+            trash.push_back(*elem);
+            continue; // skips the next elem->request(1)
+          } else {
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/flowable/FlowableOperator.h]: Nit: Don't need the `else`.

File: yarpl/include/yarpl/flowable/FlowableOperator.h [hunk 8]
```diff
@@ -749,6 +798,7 @@ class FlatMapOperator : public FlowableOperator<T, R, FlatMapOperator<T, R>> {
         R value;
         bool hasValue{false};
         folly::exception_wrapper onErrorEx{nullptr};
+        bool isTerminated{false};
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/flowable/FlowableOperator.h]: Move this just below `hasValue`.  Padding is evil.

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 6 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/Flowable.h]: That's for the using alias which is the `::value` of the type; not the type itself

Comment (dymk) [reply yarpl/include/yarpl/flowable/Flowable.h]:
```
template <typename T>
using is_t_flowable_v = is_t_flowable<T>::value
```
would be the convention iirc

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 5 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/Flowable.h]: @alexmalyshev the `enable_if` is to SFINAE out `merge` when T is not a Flowable

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 4 above)
Comment (dymk) [reply yarpl/include/yarpl/flowable/Flowable.h]: (y) fixed flatMap to deduce types from the lambda its given directly

File: yarpl/include/yarpl/flowable/Flowable.h (hunk 5 above)
Comment (alexmalyshev) [reply yarpl/include/yarpl/flowable/Flowable.h]: I know, my opinion is that SFINAE produces worse errors than a static_assert failure.

*goes off to the hills to shout about C++ Concepts*

**Comment by phoad** [comment]
P58748359 -- does your update fix this issue?

**Comment by dymk** [comment]
@phoad it does :)

File: yarpl/include/yarpl/Refcounted.h [hunk 9]
```diff
@@ -311,56 +312,49 @@ class AtomicReference final {
...
   Reference<T> load() {
-    return Reference<T>(*this);
+    std::lock_guard<std::mutex> lg{mtx_};
+    if (pointer_) {
+      pointer_->incRef();
+      Reference<T> r{pointer_, detail::skip_initial_refcount_check{}};
+      return r;
+    }
+    else {
+      Reference<T> r{nullptr};
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: This else block is equivalent to

  return Reference<T>{};

File: yarpl/include/yarpl/Refcounted.h [hunk 10]
```diff
@@ -311,56 +312,49 @@ class AtomicReference final {
...
-  AtomicReference& operator=(Reference<U>&& other) {
-    return assign(std::move(other));
-  }
-
   Reference<T> load() {
-    return Reference<T>(*this);
+    std::lock_guard<std::mutex> lg{mtx_};
+    if (pointer_) {
+      pointer_->incRef();
+      Reference<T> r{pointer_, detail::skip_initial_refcount_check{}};
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: You can get rid of `r` and return the `Reference<T>{...}` directly.

File: yarpl/include/yarpl/Refcounted.h [hunk 11]
```diff
@@ -369,12 +363,14 @@ class AtomicReference final {
...
-    if (auto p = pointer_.load()) {
-      p->decRef();
+    if (pointer_) {
+      pointer_->decRef();
     }
   }
 
-  std::atomic<T*> pointer_{nullptr};
+  std::mutex mtx_;
+  // std::atomic<T*> pointer_{nullptr};
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: Dead code?

File: yarpl/include/yarpl/Refcounted.h [hunk 12]
```diff
@@ -369,12 +363,14 @@ class AtomicReference final {
...
 
-    if (auto p = pointer_.load()) {
-      p->decRef();
+    if (pointer_) {
+      pointer_->decRef();
     }
   }
 
-  std::atomic<T*> pointer_{nullptr};
+  std::mutex mtx_;
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: sadpanda

File: yarpl/include/yarpl/flowable/Flowable.h [hunk 13]
```diff
@@ -20,6 +20,21 @@
...
+class Flowable;
+
+namespace detail {
+
+template <typename T>
+struct is_t_flowable : std::false_type {};
+
+template <typename R>
+struct is_t_flowable<Reference<Flowable<R>>> : std::true_type {
+  using ElemType = R;
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/flowable/Flowable.h]: Nit: Using underscore-lowercase for `is_t_flowable` and then upper-camelcase for `ElemType` is a little odd.

I'd suggest picking one, either `is_t_flowable::value_type` or `IsFlowable::ElemType`.

File: yarpl/include/yarpl/flowable/FlowableOperator.h [hunk 14]
```diff
@@ -93,7 +93,12 @@ class FlowableOperator : public Flowable<D> {
     // Subscriber.
 
     void onSubscribeImpl() override {
-      subscriber_->onSubscribe(this->ref_from_this(this));
+      if (auto s = subscriber_.load()) {
+        s->onSubscribe(this->ref_from_this(this));
+      }
+      else {
+        CHECK(false) << "subscriber_ is null";
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/flowable/FlowableOperator.h]: Why not just crash?  The `CHECK(false)` doesn't really get you much.
```
  auto s = subscriber_.load();
  s->onSubscribe(...);
```

Also `CHECK(false)` should pretty much be `LOG(FATAL)` always.

File: yarpl/include/yarpl/Refcounted.h [hunk 15]
```diff
@@ -311,56 +312,49 @@ class AtomicReference final {
...
     inc();
   }
 
-  explicit operator bool() const {
-    return pointer_;
-  }
-
-  T* operator->() const {
-    return pointer_;
+  void store(std::nullptr_t) {
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/Refcounted.h]: Why is this needed?  Is there not an implicit `Reference<T>` ctor from `std::nullptr_t`?

Comment (dymk) [reply yarpl/include/yarpl/Refcounted.h]: Just an optimization, same as the Reference<> case. It would also be just excellent if we could get rid of Reference and avoid all this mess.

File: yarpl/include/yarpl/flowable/FlowableOperator.h [hunk 16]
```diff
@@ -718,12 +782,19 @@ class FlatMapOperator : public FlowableOperator<T, R, FlatMapOperator<T, R>> {
           : flatMapSubscription_(std::move(subscription)) {}
 
       void onSubscribeImpl() final {
+#ifdef NDEBUG
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/flowable/FlowableOperator.h]: Did you mean `#ifndef NDEBUG`?  For this, and all follow up uses?  You want the debugging code in the cases where NDEBUG is _not defined_ right, or am I getting myself all flipped up.

File: yarpl/include/yarpl/flowable/FlowableOperator.h [hunk 17]
```diff
@@ -700,6 +757,13 @@ class FlatMapOperator : public FlowableOperator<T, R, FlatMapOperator<T, R>> {
       else {
         requested_ += n;
       }
+
+      if (n > 0) {
+        // TODO: make max parallelism configurable a-la RxJava 2.x's
+        // FlowableFlatMapOperator
+        SuperSubscription::request(std::numeric_limits<int64_t>::max());
```
Comment (alexmalyshev) [inline yarpl/include/yarpl/flowable/FlowableOperator.h]: For the short term I think this should be int32_t::max(), cause we have a lot of bad assertions in rsocket that assume that.
